Remove stream key from log statements

mv2731 opened the pull request:
Previously, both the ffmpeg and OBS log exposed the stream key in the HLS ingestion URL.
This pull request fixes the issue by doing the following:

1. Set a field called printable in the ffmpeg_muxer struct, the printable equivalent of path
2. Set a member of main_params to be print_file, the printable equivalent of file 
3. Edit the code logic surrounding sensitive prints, so that the printable version is checked first, and if set, printed. 
4. Write a logging callback function for ffmpeg and replace the stream key with "{stream_key}" if ever it is detected

Relevant issues:
https://github.com/obsproject/obs-studio/pull/3114

mv2731 committed:
obs-ffmpeg: Remove stream_key from ffmpeg logs

Remove stream_key from log statements made by ffmpeg
because ffmpeg prints channel stream_key directly to
the log screen. Replace the individual stream_keys with
the string {stream_key}. Set a callback logging function
for ffmpeg that checks each line for stream_key. Try to
mimic original formatting and color of ffmpeg log.

mv2731 committed:
obs-ffmpeg: Add color to the ffmpeg log

Add original color and spacing to the ffmpeg
log. Also add checks whether global_stream_key
is set or not to make the code more robust and
less prone to error.

mv2731 committed:
obs-ffmpeg: Apply clang-format

mv2731 committed:
obs-ffmpeg: Fix CMake spacing

mv2731 committed:
obs-ffmpeg: Clean up obs-ffmpeg-mux.c

mv2731 committed:
obs-ffmpeg: Remove spacing

mv2731 committed:
obs-ffmpeg: Remove print statements

mv2731 committed:
obs-ffmpeg: Remove unnecessary brackets

mv2731 committed:
obs-ffmpeg: Add debug messages

mv2731 committed:
obs-ffmpeg: Add in black background

mv2731 committed:
obs-ffmpeg: Add black background to warning

mv2731 committed:
obs-ffmpeg: Remove spaces

mv2731 committed:
obs-ffmpeg: Remove extraneous prints and colors

mv2731 committed:
obs-ffmpeg: Add helper functions for stream_key

mv2731 committed:
obs-ffmpeg: Refactor code to hide stream_key

mv2731 committed:
obs-ffmpeg: Ensure stream key not printed by ffmpeg

mv2731 committed:
obs-ffmpeg: Remove print statements

ushadow commented on plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c:
```diff
@@ -27,8 +27,16 @@
 #include <stdlib.h>
 #include "ffmpeg-mux.h"
 
+#include <obs-module.h>
```
Is obs-module.h needed?

ushadow commented on plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c:
```diff
@@ -70,6 +80,7 @@ static inline void resize_buf_free(struct resize_buf *rb)
 
 struct main_params {
 	char *file;
+	char *print_file;
```
printable_file and add a comment about this field.

mv2731 committed:
obs-ffmpeg: Change printable to printable_path

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-mux.c:
```diff
@@ -45,6 +45,7 @@ struct ffmpeg_muxer {
 	int64_t stop_ts;
 	uint64_t total_bytes;
 	struct dstr path;
+	const char *printable_path;
```
Also add a comment about printable_path: It's the path with stream key information removed.

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-mux.c:
```diff
@@ -420,13 +443,16 @@ static bool ffmpeg_hls_mux_start(void *data)
 
 static int deactivate(struct ffmpeg_muxer *stream, int code)
 {
+	obs_service_t *service;
+	const char *temp;
```
These two variables are not needed anymore.

mv2731 replied:
yep thanks!

mv2731 committed:
obs-ffmpeg: Change printable to printable_path

Change printable field of ffmpeg_muxer struct to
printable_path and perform the same on the print_file
field of main_params, changing it to printable_file.
Add comments explaining that these are version of path
and file respectively that do not include information
about the stream key.

mv2731 committed:
obs-ffmpeg: Commit ffmpeg-mux.c changes

ushadow commented on plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c:
```diff
@@ -257,6 +302,16 @@ static bool init_params(int *argc, char ***argv, struct main_params *params,
+		struct dstr tmp = {0};
+		dstr_copy(&tmp, params->file);
+		dstr_replace(&tmp, global_stream_key, "{stream_key}");
+		strcpy(params->file, tmp.array);
```
Shouldn't tmp.array be assigned to printable_file? I don't think HLS streaming is working with this change.

mv2731 replied:
Yes! I think this occurred when I changed the names--you're completely right. Even though the log looks fine it doesn't successfully stream to YouTube. Thanks for the catch.

mv2731 committed:
obs-ffmpeg: Correct strcpy call to printable

mv2731 committed:
obs-ffmpeg: Run clang-format

ushadow commented on plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c:
```diff
@@ -70,6 +79,8 @@ static inline void resize_buf_free(struct resize_buf *rb)
 struct main_params {
 	char *file;
+	/* printable_file is file with any stream key information removed */
+	char printable_file[4096];
```
Can struct dstr be used for printable_file?

ushadow reviewed (commented):
Please see comment.

mv2731 committed:
obs-ffmpeg: Change printable_file to a dstr

ushadow commented on plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c:
```diff
@@ -257,6 +304,16 @@ static bool init_params(int *argc, char ***argv, struct main_params *params,
+	get_opt_str(argc, argv, &global_stream_key, "stream key");
+	if (strcmp(global_stream_key, "") != 0) {
+		struct dstr tmp = {0};
+		dstr_copy(&tmp, params->file);
```
You can just directly copy to printable_file. Don't need tmp.

ushadow commented on plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c:
```diff
@@ -471,8 +528,11 @@ static inline int open_output_file(struct ffmpeg_mux *ffm)
-			fprintf(stderr, "Couldn't open '%s', %s",
-				ffm->params.file, av_err2str(ret));
+			fprintf(stderr, "Couldn't open '%s', %s\n",
+				ffm->params.printable_file.len
```
dstr_is_empty(&ffm->params.pritable_file), and same elsewhere.

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-mux.c:
```diff
@@ -217,6 +219,27 @@ static void log_muxer_params(struct ffmpeg_muxer *stream, const char *settings)
+{
+	const char *stream_key;
+
+	stream_key = get_stream_key(stream);
```
I think it's not necessary to pass stream key in the command when it is not HLS. We can store stream_key as a member field in struct ffmpeg_muxer and set only in ffmpeg_hls_mux_start.

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-mux.c:
```diff
@@ -45,6 +45,8 @@ struct ffmpeg_muxer {
 	uint64_t total_bytes;
 	struct dstr path;
+	/* printable_path is path with any stream key information removed */
+	const char *printable_path;
```
Use struct dstr for printable_path, then you don't need to worry about if the array pointed by the pointer could be deallocated or not.

ushadow reviewed (changes requested):
Please see the comments.

mv2731 committed:
obs-ffmpeg: Create ffmpeg_muxer field for printable

Create dstr fields in ffmpeg_muxer for printable_path and for
stream_key so that the object can simply be stored the first
time it is obtained.

mv2731 committed:
obs-ffmpeg: Fix compile errors

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-mux.c:
```diff
@@ -372,7 +386,7 @@ static bool ffmpeg_hls_mux_start(void *data)
 	struct ffmpeg_muxer *stream = data;
 	obs_service_t *service;
 	const char *path_str;
-	const char *stream_key;
```
Just use stream_key.

mv2731 committed:
obs-ffmpeg: Change streamkey to stream_key

ushadow commented on plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c:
```diff
@@ -200,6 +213,40 @@ static bool get_audio_params(struct audio_params *audio, int *argc,
+
+	case AV_LOG_WARNING:
+		fprintf(stdout, "warning: %s[ffmpeg_muxer]%s %s%s%s",
+			ANSI_COLOR_MAGENTA, ANSI_COLOR_RESET, ANSI_COLOR_YELLOW,
```
Just use magenta for the whole line including warning.

ushadow commented on plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c:
```diff
@@ -200,6 +213,40 @@ static bool get_audio_params(struct audio_params *audio, int *argc,
+
+	case AV_LOG_ERROR:
+		fprintf(stderr, "error: %s[ffmpeg_muxer]%s %s%s%s",
+			ANSI_COLOR_BLK, ANSI_COLOR_RESET, ANSI_COLOR_RED,
```
Just use red for the whole line including error.

ushadow reviewed (changes requested):
Please see comments.

mv2731 committed:
obs-ffmpeg: Adjust text color for white background

mv2731 committed:
obs-ffmpeg: Add color to full error line
